src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -378,92 +378,186 @@ public static bool TryFormatDouble(double value, ReadOnlySpan<char> format, Numb
         {
-            char fmt = ParseFormatSpecifier(format, out int digits);
-            int precision = DoublePrecision;
+            if (fmt == 0)
+            {
+                isSignificantDigits = true;
+                return precision;
+            }
 
-            byte* pDigits = stackalloc byte[DoubleNumberBufferLength];
-            NumberBuffer number = new NumberBuffer(NumberBufferKind.FloatingPoint, pDigits, DoubleNumberBufferLength);
+            int maxDigits = precision;
 
             switch (fmt)
             {
-                case 'R':
-                case 'r':
+                case 'C':
+                case 'c':
                 {
-                    // In order to give numbers that are both friendly to display and round-trippable, we parse the
-                    // number using 15 digits and then determine if it round trips to the same value. If it does, we
-                    // convert that NUMBER to a string, otherwise we reparse using 17 digits and display that.
-                    DoubleToNumber(value, DoublePrecision, ref number);
+                    // The currency format uses the precision specifier to indicate the number of
+                    // decimal digits to format. This defaults to NumberFormatInfo.CurrencyDecimalDigits.
 
-                    if (number.Scale == ScaleNAN)
-                    {
-                        return info.NaNSymbol;
-                    }
-                    else if (number.Scale == ScaleINF)
+                    if (precision == -1)
                     {
-                        return number.IsNegative ? info.NegativeInfinitySymbol : info.PositiveInfinitySymbol;
+                        precision = info.CurrencyDecimalDigits;
                     }
+                    isSignificantDigits = false;
 
-                    double roundTrip = NumberToDouble(ref number);
+                    break;
+                }
 
-                    if (roundTrip == value)
-                    {
-                        NumberToString(ref sb, ref number, 'G', DoublePrecision, info);
-                    }
-                    else
+                case 'E':
+                case 'e':
+                {
+                    // The exponential format uses the precision specifier to indicate the number of
+                    // decimal digits to format. This defaults to 6. However, the exponential format
+                    // also always formats a single integral digit, so we need to increase the precision
+                    // specifier and treat it as the number of significant digits to account for this.
+
+                    if (precision == -1)
                     {
-                        DoubleToNumber(value, 17, ref number);
-                        NumberToString(ref sb, ref number, 'G', 17, info);
+                        precision = 6;
```

Comment:
Use named constant?

Reply:
Will fix (and the other existing consumer of this value in `NumberToString`).